### PR TITLE
Changes to mxnet.metric

### PR DESCRIPTION
**Pull request by acphile**
## Description ##
change based on https://github.com/apache/incubator-mxnet/issues/18046 
1. make improvements in metric
  a. improve Class MAE (and MSE, RMSE)
  b. improve Class _BinaryClassification
  c. improve Class TopKAccuracy
  d. add Class MeanCosineSimilarity
  e. add Class MeanPairwiseDistance
2. move mxnet.metric to mxnet.gluon.metric

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/gluon/metric.py [hunk 1]
```diff
@@ -777,16 +815,113 @@ class F1(EvalMetric):
...
+
+        preds : list of `NDArray`
+            Predicted values.
+        """
+        labels, preds = check_label_shapes(labels, preds, True)
+
+        for label, pred in zip(labels, preds):
+            self.metrics.update_binary_stats(label, pred)
+
+        if self.average == "macro":
```
Comment (sxjscience) [inline python/mxnet/gluon/metric.py:848]: In fact, `macro averaging` + F1 does not mean to average the F1 of each batch. I think we should revise it to be the same as https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html .

File: python/mxnet/gluon/metric.py [hunk 2]
```diff
@@ -1130,12 +1339,18 @@ def update(self, labels, preds):
                 label = label.reshape(label.shape[0], 1)
             if len(pred.shape) == 1:
                 pred = pred.reshape(pred.shape[0], 1)
-
-            mae = numpy.abs(label - pred).mean()
+            
+            if self.average == "macro":
```
Comment (sxjscience) [inline python/mxnet/gluon/metric.py:1343]: It's very strange to have `macro` + MAE. See scikit-learn: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_absolute_error.html#sklearn.metrics.mean_absolute_error

File: python/mxnet/gluon/metric.py (hunk 1 above)
Comment (acphile) [reply python/mxnet/gluon/metric.py:848]: Averaging F1 per batch here previously existed in metric.py before I made changes. This calculation also exists in MAE, MSE, RMSE, and PearsonCorrelation. Should I remove all of them accordingly? For the average "macro" in sklearn, it seems used in calculating F1 score for multiclass/multilabel targets. But currently our F1 only supports binary classification. I think I need to make extensions for F1.

Comment (sxjscience) [reply python/mxnet/gluon/metric.py:848]: Let's remove it and make it similar to sklearn. This is in fact the reason why I never use the metric class in MXNet.

File: python/mxnet/gluon/metric.py [hunk 3]
```diff
@@ -619,90 +583,100 @@ def update_binary_stats(self, label, pred):
...
     @property
     def fscore(self):
-        if self.precision + self.recall > 0:
-            return 2 * self.precision * self.recall / (self.precision + self.recall)
-        else:
-            return 0.
+        return (1 + self.beta ** 2) * self.precision * self.recall / numpy.maximum(self.beta ** 2 * self.precision + self.recall, 1e-12)
 
     @property
     def global_fscore(self):
```
Comment (leezu) [inline python/mxnet/gluon/metric.py:664]: This method should be removed as you dropped the global states?

File: python/mxnet/gluon/metric.py [hunk 4]
```diff
@@ -24,9 +24,9 @@
 
 import numpy
```
Comment (leezu) [inline python/mxnet/gluon/metric.py:25]: Instead of using numpy, we can use `mxnet.numpy` as it runs asynchronously and has GPU support.
The summary states of a metric should be stored on CPU, but if for example data and label are on GPU and input to the metric, we can calculate the sufficient statistics on GPU

File: python/mxnet/gluon/metric.py (hunk 3 above)
Comment (acphile) [reply python/mxnet/gluon/metric.py:664]: This method actually refers to the micro calculation for F1 and it is not related to original global support.

Comment (leezu) [reply python/mxnet/gluon/metric.py:664]: Would it make sense to adjust the name?

Comment (acphile) [reply python/mxnet/gluon/metric.py:664]: I think it is ok to use global_fscore since it is in a private container class.

File: python/mxnet/gluon/metric.py [hunk 5]
```diff
@@ -617,92 +585,102 @@ def update_binary_stats(self, label, pred):
...
-
-        check_label_shapes(label, pred)
-        if len(numpy.unique(label)) > 2:
-            raise ValueError("%s currently only supports binary classification."
-                             % self.__class__.__name__)
+        pred = pred.as_np_ndarray()
+        label = label.as_np_ndarray().astype('int32')
+        if self.class_type == "binary":
+            self._set(1)
+            if len(numpy.unique(label)) > 2:
```
Comment (leezu) [inline python/mxnet/gluon/metric.py:592]: This will trigger synchronization (as we need to wait for the result of the np.unique operator). Could we make error checking that triggers synchronization optional?

Comment (leezu) [reply python/mxnet/gluon/metric.py:592]: https://github.com/apache/incubator-mxnet/blob/83b51703ed354f41024423f140de38df2ba22d50/src/imperative/imperative.cc#L123-L127

**Comment by acphile** [comment]
@mxnet-bot  run ci [centos-cpu, sanity, centos-gpu, windows-cpu, unix-gpu, windows-gpu, unix-cpu]

**Comment by leezu** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by leezu** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by leezu** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by marcoabreu** [comment]
Why are we removing tvm op?

**Comment by leezu** [comment]
It's currently broken. https://github.com/apache/incubator-mxnet/issues/17840 Needs to be fixed before it can be enabled on CI

**Comment by marcoabreu** [comment]
Please disable it in a separate PR then. I'd like to have a better understanding of what exactly is broken , but not have that discussion in this PR.

**Comment by leezu** [comment]
Let's disable it here, because it blocks this PR

**Comment by leezu** [comment]
You can go ahead and try reproduce the issue locally:

The reproducer is available since more than a month at https://github.com/apache/incubator-mxnet/issues/17886#issue-585864921

**Comment by marcoabreu** [comment]
No, it's unrelated and should be a separated and isolated PR. Each PR should serve one propose. That way, we can focus discussions, have single purpose commits and also allow reverting

**Comment by leezu** [comment]
You're very welcome to resolve this issue and create a separate PR

**Comment by leezu** [comment]
Let's disable it in https://github.com/apache/incubator-mxnet/pull/18204

**Comment by acphile** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by acphile** [comment]
@mxnet-bot run ci [unix-cpu, windows-gpu]

File: tests/python/tensorrt/test_cvnets.py [hunk 6]
```diff
@@ -29,7 +28,12 @@
 def get_classif_model(model_name, use_tensorrt, ctx=mx.gpu(0), batch_size=128):
     mx.contrib.tensorrt.set_use_fp16(False)
     h, w = 32, 32
-    net = gluoncv.model_zoo.get_model(model_name, pretrained=True)
+    model_url = "https://raw.githubusercontent.com/dmlc/web-data/master/gluoncv/models/"
```
Comment (leezu) [inline tests/python/tensorrt/test_cvnets.py:31]: Please don't hardcode `master` in the URL here. The repository may change and will then break the CI. Instead, use the commit ID: `https://raw.githubusercontent.com/dmlc/web-data/221ce5b7c6d5b0777a1e3471f7f03ff98da90a0a/gluoncv/models/"

**Comment by acphile** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by acphile** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by mseth10** [comment]
@acphile this PR fails nightly CD while running nightly python unit tests. The following tests fail:
test_mcc, test_multilabel_f1, test_binary_f1
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/restricted-mxnet-cd%2Fmxnet-cd-release-job/detail/mxnet-cd-release-job/1119/pipeline/329

We'll need to revert this PR and fix the failures before re-merging it.

**Comment by leezu** [comment]
The reason it fails is that the CI check in this PR was run too long ago. I should have restarted it before merging the PR. Meantime master changed and caused some additional changes to be necessary. They are in https://github.com/apache/incubator-mxnet/pull/18312/files
